src/make-gecko-action-hooks.js

```diff
@@ -0,0 +1,155 @@
+const chalk = require('chalk');
+
+module.exports.setup = (program) => {
+  return program
+    .command('make-gecko-action-hooks')
+    .option('-n, --noop', 'Don\'t change roles, just show difference')
+    .description('create or update gecko in-tree action hooks');
+};
+
+module.exports.run = async function(options) {
+  let taskcluster = require('taskcluster-client');
+  let chalk = require('chalk');
+  let projects = await getProjects();
+
+  // We build action hooks' task definitions from the latest in-tree `.taskcluster.yml`.
+  let taskclusterYml = await getCentralTaskclusterYml();
+
+  for (let action of ACTION_HOOKS) {
+    const hookGroupId = `project-${action.trustDomain}`;
+    const hookId = `in-tree-action-${action.level}-${action.actionPerm}`;
+    const {task, triggerSchema} = makeHookDetails(taskclusterYml, action);
+    await editHook({
+      noop: options.noop,
+      hookGroupId,
+      hookId,
+      metadata: {
+        name: `Action task ${action.level}-${action.actionPerm}`,
+        description: [
+          '*DO NOT EDIT*',
+          '',
+          'This hook is configured automatically by',
+          '[taskcluster-admin](https://github.com/taskcluster/taskcluster-admin).',
+        ].join('\n'),
+        owner: 'taskcluster-notifications@mozilla.com',
+        emailOnError: false, // true, TODO
+      },
+      schedule: [],
+      task,
+      triggerSchema,
+    });
+
+    // make the role with scopes assume:repo:<repo>:action:<actionPerm> for each repo at this level
+    const projectsAtLevel = Object.keys(projects)
+      .filter(p => projects[p].access === `scm_level_${action.level}`)
+      .map(p => projects[p]);
+    const scopes = projectsAtLevel.map(
+      project => `assume:repo:hg.mozilla.org/${hgmoPath(project)}:action:${action.actionPerm}`);
+    await editRole({
+      roleId: `hook-id:${hookGroupId}/${hookId}`,
+      description: [
+        '*DO NOT EDIT*',
+        '',
+        'This role is configured automatically by',
+        '[taskcluster-admin](https://github.com/taskcluster/taskcluster-admin).',
+      ].join('\n'),
+      scopes,
+      noop: options.noop,
+    });
+  }
+};
+
+const makeHookDetails = (taskclusterYml, action) => {
+  const repoUrlExpression = '${payload.decision.repository.url[8:]}';
+  const actionOverrides = {
+    repo_scope: {
+      $let: {
+        repoUrl: {
+          $if: 'payload.decision.repository.url[-1] == "/"',
+          then: {$eval: 'payload.decision.repository.url[:-1]'},
+          else: {$eval: 'payload.decision.repository.url'},
+        },
+      },
+      in: 'assume:repo:${repoUrl[8:]}:action:' + action.actionPerm,
+    },
+  };
+
+  if (action.actionPerm !== 'generic') {
+    // enforce that action.cb_name == actionPerm
+    actionOverrides.cb_name = action.actionPerm;
+  }
+
+  const task = {
+    $let: {
+      tasks_for: 'action',
+      action: {
+        $merge: [
+          {$eval: 'payload.decision.action'},
```

Comment:
There are only two, and one of them (cb_name) is only sometimes overridden.  So I'm not sure that'd be less confusing.  Also, a sub-key would still require merging..

Reply:
My concern is less about what the code looks like than about security; making it harder to future changes to accidentally let the security sensitive values be overridden (which suggests the name should be something like `untrusted`, maybe).

To be clear, I don't think the code as-is is insecure. I just think that it would be easy for somebody to not realize the security impact of the code (for example, that the order of merging is critical). By moving the security critical values outside of the user-provided dictionary, I think that makes mistakes here less likely.

Reply:
Hm, I see.  So, a few things to change:
 * add detail to hooks' `triggerSchema` to indiciate specific keys in action, push, and repository
 * don't include `repo_scope` in the hookPayload (and don't generate it in the in-tree code)
 * Rework these overrides to be a little clearer that we either take the given value, or force it, and why (with some comments).